src/utils/prometheus.ts

```diff
@@ -0,0 +1,183 @@
+      const [onchainNonce, engineNonce] = await Promise.all([
+        getLastUsedOnchainNonce(chainId, walletAddress),
+        inspectNonce(chainId, walletAddress),
+      ]);
+
+      this.set(
+        {
+          wallet: walletAddress,
+          chain: chainId.toString(),
+          type: "onchain",
+        },
+        onchainNonce,
+      );
+      this.set(
+        {
+          wallet: walletAddress,
+          chain: chainId.toString(),
+          type: "engine",
+        },
+        engineNonce,
+      );
+    }
+  },
+});
+
+export const register = new Registry();
+
+register.registerMetric(nonceMetrics);
+
+// Define strict types for each event
+type ResponseSentParams = {
+  endpoint: string;
+  statusCode: string;
+  method: string;
+  duration: number;
+};
+
+type TransactionQueuedParams = {
+  chainId: string;
+  walletAddress: string;
+};
+
+type TransactionSentParams = TransactionQueuedParams & {
+  success: boolean;
+  duration: number;
+};
+
+type TransactionMinedParams = TransactionQueuedParams & {
+  endToEndDuration: number;
+  duration: number;
+};
+
+// Union type for all possible event parameters
+type MetricParams =
+  | { event: "response_sent"; params: ResponseSentParams }
+  | { event: "transaction_queued"; params: TransactionQueuedParams }
+  | { event: "transaction_sent"; params: TransactionSentParams }
+  | { event: "transaction_mined"; params: TransactionMinedParams };
+
+// Request metrics
+const requestsTotal = new Counter({
+  name: "engine_requests_total",
+  help: "Total number of completed requests",
+  labelNames: ["endpoint", "status_code", "method"],
+  registers: [register],
+});
+
+const requestDuration = new Histogram({
+  name: "engine_response_time_ms",
+  help: "Response time in milliseconds",
+  labelNames: ["endpoint", "status_code"],
+  buckets: [100, 300, 500, 700, 1000, 3000, 5000, 7000, 10_000],
+  registers: [register],
+});
+
+// Transaction metrics
+const transactionsQueued = new Gauge({
+  name: "engine_transactions_queued",
+  help: "Number of transactions currently in queue",
+  labelNames: ["chain_id", "wallet_address"],
+  registers: [register],
+});
+
+const transactionsQueuedTotal = new Counter({
+  name: "engine_transactions_queued_total",
+  help: "Total number of transactions queued",
+  labelNames: ["chain_id", "wallet_address"],
+  registers: [register],
+});
+
+const transactionsSentTotal = new Counter({
+  name: "engine_transactions_sent_total",
+  help: "Total number of transactions sent",
+  labelNames: ["chain_id", "wallet_address", "is_success"],
+  registers: [register],
+});
+
+const transactionsMinedTotal = new Counter({
+  name: "engine_transactions_mined_total",
+  help: "Total number of transactions mined",
+  labelNames: ["chain_id", "wallet_address"],
+  registers: [register],
+});
+
+// Transaction duration histograms
+const queueToSentDuration = new Histogram({
+  name: "engine_transaction_queue_to_sent_seconds",
```

Comment:
we should pick `engine_transactions_` or `engine_transaction_` 
singular or plural, and be consistent

Reply:
traefik does this:
for histograms: `traefik_service_request_duration_seconds_sum`
for counters: `traefik_service_requests_total`